### PR TITLE
feat(mtls): trust store and mtls

### PR DESCRIPTION
**Pull request by jonahkaye**
Refs: #[1377](https://github.com/metriport/metriport-internal/issues/1377)

### Dependencies

- Upstream: https://github.com/metriport/metriport-internal/pull/1435

### Description

- implementing mTLS for API GW. This required a refactor to use aws-apigatewayv2 which supports adding an ownershipCertificate to the domain name, which is needed if you are running mTLS with private certs. See cdk [docs](https://docs.aws.amazon.com/cdk/api/v2/docs/aws-cdk-lib.aws_apigatewayv2.HttpApi.html) for implementation I followed. Basically we needed to use a HTTP api instead of a REST API
- added truststore bucket and bundle: https://s3.console.aws.amazon.com/s3/buckets/mtls-trust-bundle-staging?region=us-west-1&bucketType=general&tab=objects
- got an OwnershipCertificate issued by AWS to ihe.staging.metriport.com: https://us-east-2.console.aws.amazon.com/acm/home?region=us-east-2#/certificates/745df350-1970-41cc-bcf9-8c23977a441d
- discovered tha WAF is not compatible with HTTP api 
- Because this PR required using the same A record and Domain name in the new gateway as the old one, it actually had a more complicated deployment process. Needed to branch to staging with all the IHE stack code basically commented out, and then push changes, and then comment it out again to restore staging to a clean state

### Testing

- Staging
  - [x] branch to staging
  - [x] test there with our own cert
  - [x] verify that without the cert we cant get throug
  - [ ] have testing partner ping us (qvera). Or we could just get their certs from their endpoints and test ourselves hehe. That saves effort since I dont have to mock responses or wait on them so they dont freak out

### Release Plan

- [ ] Merge this. Note that in order for this to not fail on release to staging, will need to branch to staging first with a branch that removes IHE stack code, and then merge. See above for why. This should not be required when ship to prod since there is no IHE stack on prod yet 
- [x] add to go Live that we need to add counterpart bucket and bundle to aws

### CODE REVIEW

File: packages/lambdas/src/patient-discovery.ts [hunk 1]
```diff
@@ -17,6 +17,7 @@ export const handler = Sentry.AWSLambda.wrapHandler(async (event: APIGatewayProx
     id: payload.id,
     timestamp: payload.timestamp,
     samlAttributes: payload.samlAttributes,
+    patientResource: payload.patientResource,
```
Comment (jonahkaye) [inline packages/lambdas/src/patient-discovery.ts:20]: tiny improvement to pd parsing

File: packages/infra/lib/ihe-stack.ts [hunk 2]
```diff
@@ -39,92 +42,129 @@ export class IHEStack extends Stack {
...
-
-    // add domain cert + record
-    const iheApiUrl = `${props.config.iheGateway?.subdomain}.${props.config.domain}`;
-    api.addDomainName("IHEAPIDomain", {
-      domainName: iheApiUrl,
-      certificate: certificate,
-      securityPolicy: apig.SecurityPolicy.TLS_1_2,
-    });
-    new r53.ARecord(this, "IHEAPIDomainRecord", {
+    // commenting out so CFN deletes
```
Comment (RamilGaripov) [inline packages/infra/lib/ihe-stack.ts:99]: left-over comment?

**Review by RamilGaripov — APPROVED** [review]
LGTM!

**Review by jonahkaye — COMMENTED** [review]
mTLS working with client providing a cert and verifying server cert: 

```
curl --cert ihe.staging.metriport.com.pem --key DECRYPTEDKEY.key --cacert root_cert.pem https://ihe.staging.metriport.com/ -v
*   Trying 52.14.212.22:443...
* Connected to [ihe.staging.metriport.com](http://ihe.staging.metriport.com/) (52.14.212.22) port 443 (#0)
* ALPN: offers h2,http/1.1
* (304) (OUT), TLS handshake, Client hello (1):
*  CAfile: root_cert.pem
*  CApath: none
* (304) (IN), TLS handshake, Server hello (2):
* TLSv1.2 (IN), TLS handshake, Certificate (11):
* TLSv1.2 (IN), TLS handshake, Server key exchange (12):
* TLSv1.2 (IN), TLS handshake, Request CERT (13):
* TLSv1.2 (IN), TLS handshake, Server finished (14):
* TLSv1.2 (OUT), TLS handshake, Certificate (11):
* TLSv1.2 (OUT), TLS handshake, Client key exchange (16):
* TLSv1.2 (OUT), TLS handshake, CERT verify (15):
* TLSv1.2 (OUT), TLS change cipher, Change cipher spec (1):
* TLSv1.2 (OUT), TLS handshake, Finished (20):
* TLSv1.2 (IN), TLS change cipher, Change cipher spec (1):
* TLSv1.2 (IN), TLS handshake, Finished (20):
* SSL connection using TLSv1.2 / ECDHE-RSA-AES128-GCM-SHA256
* ALPN: server accepted h2
* Server certificate:
*  subject: C=US; ST=California; O=Metriport Inc.; OU=CAREQUALITY-TEST; CN=[ihe.staging.metriport.com](http://ihe.staging.metriport.com/)
*  start date: Nov  2 21:25:42 2023 GMT
*  expire date: Nov  2 21:25:42 2024 GMT
*  subjectAltName: host "[ihe.staging.metriport.com](http://ihe.staging.metriport.com/)" matched cert's "[ihe.staging.metriport.com](http://ihe.staging.metriport.com/)"
*  issuer: C=US; ST=California; L=San Diego; O=EMR Direct; OU=Certification Authority ([certs.emrdirect.com](http://certs.emrdirect.com/)); CN=EMR Direct Test Device SubCA2
*  SSL certificate verify ok.
```

**Review by jonahkaye — COMMENTED** [review]
If client doesnt provide a cert or cert is invalid, server resets connection:

```
curl --cacert root_cert.pem https://ihe.staging.metriport.com/pd -v
*   Trying 3.131.27.39:443...
* Connected to [ihe.staging.metriport.com](http://ihe.staging.metriport.com/) (3.131.27.39) port 443 (#0)
* ALPN: offers h2,http/1.1
* (304) (OUT), TLS handshake, Client hello (1):
*  CAfile: root_cert.pem
*  CApath: none
* (304) (IN), TLS handshake, Server hello (2):
* TLSv1.2 (IN), TLS handshake, Certificate (11):
* TLSv1.2 (IN), TLS handshake, Server key exchange (12):
* TLSv1.2 (IN), TLS handshake, Request CERT (13):
* TLSv1.2 (IN), TLS handshake, Server finished (14):
* TLSv1.2 (OUT), TLS handshake, Certificate (11):
* TLSv1.2 (OUT), TLS handshake, Client key exchange (16):
* TLSv1.2 (OUT), TLS change cipher, Change cipher spec (1):
* TLSv1.2 (OUT), TLS handshake, Finished (20):
* Recv failure: Connection reset by peer
* LibreSSL SSL_connect: Connection reset by peer in connection to ihe.staging.metriport.com:443
* Closing connection 0
curl: (35) Recv failure: Connection reset by peer
```

File: .github/workflows/_deploy_cdk.yml [hunk 3]
```diff
@@ -174,7 +174,7 @@ jobs:
         uses: metriport/deploy-with-cdk@master
         with:
           cdk_action: "diff"
-          cdk_version: "2.87.0"
+          cdk_version: "2.122.0"
```
Comment (Goncharo) [inline .github/workflows/_deploy_cdk.yml:177]: ⚠️ just wanted to flag that bumping CDK versions sometimes breaks things, so a head up incase something in the deploy goes sour that we didn't expect

File: packages/infra/lib/ihe-stack.ts [hunk 4]
```diff
@@ -39,92 +42,128 @@ export class IHEStack extends Stack {
...
+      "OwnershipVerificationCertificate",
+      props.config.iheGateway.ownershipVerificationCertArn
+    );
+
+    // get the truststore bucket from S3
+    const bucket = s3.Bucket.fromBucketName(
+      this,
+      "TruststoreBucket",
+      props.config.iheGateway?.trustStoreBucketName
+    );
```
Comment (Goncharo) [inline packages/infra/lib/ihe-stack.ts:64]: where is this bucket being created?

if we created this bucket manually, we'll need to fix it - we should only create resources manually on AWS only if absolutely necessary

File: packages/infra/config/env-config.ts [hunk 5]
```diff
@@ -98,8 +98,11 @@ export type EnvConfig = {
   iheGateway?: {
     vpcId: string;
     certArn: string;
+    ownershipVerificationCertArn: string;
     subdomain: string; // Subdomain for IHE integrations
     snsTopicArn?: string;
+    trustStoreBucketName: string;
+    trustStoreKey: string;
```
Comment (Goncharo) [inline packages/infra/config/env-config.ts:105]: the internal PR will need to be merged/approved before we can merge this one: https://github.com/metriport/metriport-internal/pull/1435/files

please update the release plan accordingly

**Review by Goncharo — CHANGES_REQUESTED** [review]
flagging this a needs review, for bookkeeping due to this: https://github.com/metriport/metriport/pull/1430#discussion_r1475293060

File: packages/infra/lib/ihe-stack.ts (hunk 4 above)
Comment (jonahkaye) [reply packages/infra/lib/ihe-stack.ts]: Issue here is I tried to create it via cdk, but then deployment fails since the trust bundle isnt inside the bucket. Chicken and egg issue but I think it justifies leaving the bucket manually created for now

File: packages/lambdas/src/inbound-patient-discovery.ts [hunk 6]
```diff
@@ -17,6 +17,7 @@ export const handler = Sentry.AWSLambda.wrapHandler(async (event: APIGatewayProx
     id: payload.id,
     timestamp: payload.timestamp,
     samlAttributes: payload.samlAttributes,
+    patientResource: payload.patientResource,
```
Comment (Goncharo) [inline packages/lambdas/src/inbound-patient-discovery.ts:20]: what does this have to do with mTLS?

File: packages/infra/lib/ihe-stack/ihe-gateway.ts [hunk 7]
```diff
@@ -84,20 +85,18 @@ export function createIHEGateway(stack: Construct, props: IHEGatewayProps): void
...
-  proxyToServer(patientDiscoveryResource, buildInboundAddress(patientDiscoveryPort));
-
-  const documentQueryResource = apiResource.addResource("document-query");
-  proxyToServer(documentQueryResource, buildInboundAddress(documentQueryPort));
-
-  const documentRetrievalResource = apiResource.addResource("document-retrieval");
-  proxyToServer(documentRetrievalResource, buildInboundAddress(documentRetrievalPort));
+  addProxyRoute("/v1/patient-discovery", buildInboundAddress(patientDiscoveryPort), apiGateway);
+  addProxyRoute("/v1/document-query", buildInboundAddress(documentQueryPort), apiGateway);
+  addProxyRoute("/v1/document-retrieval", buildInboundAddress(documentRetrievalPort), apiGateway);
```
Comment (Goncharo) [inline packages/infra/lib/ihe-stack/ihe-gateway.ts:90]: why did you remove `const apiRootV1 = api.root.addResource("v1");` in favor of putting `/v1` in every URL?

why not just build there from the `apiRootV1`?

**Review by Goncharo — COMMENTED** [review]
added some Qs - also once this PR is approved I'd suggest just creating the prod resources as well, rather than waiting for go-live ( so we don't forget)

File: packages/lambdas/src/inbound-patient-discovery.ts (hunk 6 above)
Comment (jonahkaye) [reply packages/lambdas/src/inbound-patient-discovery.ts:20]: ah nothing was just an improvement to the PD logic. Was a lingering change I just added herer

File: packages/infra/lib/ihe-stack/ihe-gateway.ts (hunk 7 above)
Comment (jonahkaye) [reply packages/infra/lib/ihe-stack/ihe-gateway.ts:90]: APIGatewayV2 doesnt support ".root" or the type that apiRoot was (an apig.IResource). This was the only way I could figure out to from the aws docs

**Review by Goncharo — APPROVED** [review]
:shipit:
